### PR TITLE
[WIP]  Disable support for scripting directives when compiling script files

### PR DESCRIPTION
**Pull request by KevinRansom**
So ... 

this disables support for scripting directives #r, #load, #I and #time in the compiler when compiling projects (rather than when compiling scripts)

This disables support for script compilation to .exe in the coreclr, and adds warnings when compiling scripts using the desktop compiler.

The warning looks like this: (the text will need some polishing, it is not yet correct)
````
foo.fsx(1,1): warning FS0076: This directive may only be used in F# script files (extensions .fsx or .fsscript). Either remove the directive, move this code to a script file or delimit the directive with '#if INTERACTIVE'/'#endif'.
````
Putting the directives in an #if INTERACTIVE region will eliminate the warning.  as will --nowarn:76

We are taking this action, primarily because, having IDE support for design-time addition and removal of references when typing, (I.e on every character) when editing projects would be performance prohibitive.

* Todo: additional testing
* fixing up existing tests

We will of course still have live reference update in the IDE for loose script files.

Kevin

### CODE REVIEW

**Comment by NinoFloris** [comment]
What's the difference on IDE perf impact vs a proj file change? Compiler could only call #r providers after closing " right?

**Comment by dsyme** [comment]
> What's the difference on IDE perf impact vs a proj file change? Compiler could only call #r providers after closing " right?

The change is only within _projects_, not within script editing.  Currently the compiler actually notices `#r` even when compiling projects (the IDE doesn't pay any attention to them)

File: src/fsharp/CompileOps.fs [hunk 1]
```diff
@@ -5098,7 +5117,7 @@ module private ScriptPreprocessClosure =
             reduceMemoryUsage) =  
 
         let projectDir = Path.GetDirectoryName(filename)
-        let isInteractive = (codeContext = CodeContext.CompilationAndEvaluation)
+        let isInteractive = true                            //(codeContext = CodeContext.CompilationAndEvaluation)
```
Comment (dsyme) [inline src/fsharp/CompileOps.fs]: What's this change?

**Comment by matthid** [comment]
> This disables support for script compilation to .exe in the coreclr, and adds warnings when compiling scripts using the desktop compiler.

So this means you can no longer compile '.fsx' files? Does this impact fcs as well? If yes, I'm afraid this will break the current FAKE design badly...

**Comment by dsyme** [comment]
> So this means you can no longer compile '.fsx' files? Does this impact fcs as well? If yes, I'm afraid this will break the current FAKE design badly...

No, it's not that.  

The thing is that the compiler was taking into account `#r` directives even when building projects (i.e. only .fs files), not scripts.  The IDE was giving errors when presented with these, but the compiler was still noticing them.  We are deprecating that as it was never the intended behavior.

**Comment by dsyme** [comment]
@KevinRansom Are we 

1. disabling the directives in `.fs` files or
2. disabling the directives when doing project compilation (and continuing to allow in `.fs` files when using `#load` from a `.fsx` file that is being compiled).

thanks

**Comment by matthid** [comment]
@dsyme thanks, I just wanted to clarify as I understood the sentence differently. While "1." would be breaking, I don't think it would be as bad as disabling compiling `.fsx` files to `.exe` or `.dll` (for netcoreapp)

**Comment by dsyme** [comment]
@KevinRansom Hmmm, we'd better work with @matthid to double check his scenario.

@matthid So you are compiling .fsx to .exe using the compiler that comes as part of the .NET Core SDK?  How are those .exe actually run?  Are they .NET Framework .exe or .NET Core .exe?  (.NET Core .exe don't really exist as a thing, right?)

**Comment by matthid** [comment]
> So you are compiling .fsx to .exe using the compiler that comes as part of the .NET Core SDK?

No. Fake is using the `netstandard` version of `FSharp.Compiler.Service`, however that should be the same code-base, correct?

>  How are those .exe actually run? Are they .NET Framework .exe or .NET Core .exe?

Technically, we [compile](https://github.com/fsharp/FAKE/blob/00d5bb2d3644c80314630591d66ed302d1b45032/src/app/Fake.Runtime/CompileRunner.fs#L104) the `fsx` as `netstandard2.0` dll and then [load it, find the main method and invoke it](https://github.com/fsharp/FAKE/blob/00d5bb2d3644c80314630591d66ed302d1b45032/src/app/Fake.Runtime/CompileRunner.fs#L49-L76). This has the huge advantage that running the "cached" version or "invoking" the script do not have different semantics (as it was in Fake 4, where people reported bugs that stuff wasn't working when running from cache) and as a bonus it works around missing APIs in netcore.

My fear is that if we disable this "feature" than I don't really know any way on how to make it work at all. The caching feature is really important for certain applications of Fake.

**Comment by KevinRansom** [comment]
1.

**Pull request closed without merging**